cmd/runtimetest/main.go

```diff
@@ -410,201 +578,379 @@ func validateDefaultFS(spec *rspec.Spec, t *tap.T) error {
+		c.harness.Skip(1, fmt.Sprintf("%s has unconfigured permissions", description))
+	} else {
+		expectedPerm := *device.FileMode & os.ModePerm
+		actualPerm := fi.Mode() & os.ModePerm
+		rfcError, err = c.Ok(actualPerm == expectedPerm, specerror.DevicesAvailable, version, fmt.Sprintf("%s has the expected permissions", description))
+		if err != nil {
+			return err
 		}
-		if device.GID != nil {
-			if *device.GID != fStat.Gid {
-				return fmt.Errorf("%v uid expected is %v, actual is %v", device.Path, *device.GID, fStat.Gid)
-			}
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"path":      device.Path,
+			"expected":  expectedPerm,
+			"actual":    actualPerm,
+		})
+	}
+
+	if description == "/dev/console (default device)" {
+		c.harness.Todo().Fail("we need the major/minor from the controlling TTY")
+		return nil
+	}
+
+	if device.UID == nil {
+		c.harness.Skip(1, fmt.Sprintf("%s has an unconfigured user ID", description))
+	} else {
+		rfcError, err = c.Ok(fStat.Uid == *device.UID, specerror.DevicesAvailable, version, fmt.Sprintf("%s has the expected user ID", description))
+		if err != nil {
+			return err
+		}
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"path":      device.Path,
+			"expected":  *device.UID,
+			"actual":    fStat.Uid,
+		})
+	}
+
+	if device.GID == nil {
+		c.harness.Skip(1, fmt.Sprintf("%s has an unconfigured group ID", description))
+	} else {
+		rfcError, err = c.Ok(fStat.Gid == *device.GID, specerror.DevicesAvailable, version, fmt.Sprintf("%s has the expected group ID", description))
+		if err != nil {
+			return err
 		}
+		c.harness.YAML(map[string]interface{}{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"path":      device.Path,
+			"expected":  *device.GID,
+			"actual":    fStat.Gid,
+		})
 	}
 
 	return nil
 }
 
-func validateDefaultSymlinks(spec *rspec.Spec, t *tap.T) error {
+func (c *complianceTester) validateDefaultSymlinks(spec *rspec.Spec) error {
 	for symlink, dest := range defaultSymlinks {
+		var exists bool
 		fi, err := os.Lstat(symlink)
+		if os.IsNotExist(err) {
+			exists = false
+		} else if err != nil {
+			return err
+		} else {
+			exists = true
+		}
+		rfcError, err := c.Ok(exists, specerror.DefaultRuntimeLinuxSymlinks, spec.Version, fmt.Sprintf("has a file at default symlink path %q", symlink))
 		if err != nil {
 			return err
 		}
-		if fi.Mode()&os.ModeSymlink != os.ModeSymlink {
-			return specerror.NewError(specerror.DefaultRuntimeLinuxSymlinks,
-				fmt.Errorf("%v is not a symbolic link as expected", symlink),
-				rspec.Version)
+		c.harness.YAML(map[string]string{
+			"level":     rfcError.Level.String(),
+			"reference": rfcError.Reference,
+			"path":      symlink,
+		})
+		if !exists {
+			continue
+		}
+
+		isSymlink := fi.Mode()&os.ModeSymlink == os.ModeSymlink
```

Comment:
I find it more clear written this way:
```
isSymlink := fi.Mode()&os.ModeType == os.ModeSymlink
```
but that's equivalent in Golang because os.ModeType is a bitfield.

It confused me at first because in C, `stat.st_mode` is not a bitfield (man 7 inode):
`S_IFLNK = S_IFREG | S_IFCHR`
```
       The following mask values are defined for the file type:

           S_IFMT     0170000   bit mask for the file type bit field

           S_IFSOCK   0140000   socket
           S_IFLNK    0120000   symbolic link
           S_IFREG    0100000   regular file
           S_IFBLK    0060000   block device
           S_IFDIR    0040000   directory
           S_IFCHR    0020000   character device
           S_IFIFO    0010000   FIFO
```

(Things I learnt today)

Reply:
I see it was already written like that before. Feel free to ignore my comment ;)